### ISSUE TITLE
Tricks to train better? I am curious as to how you got .9 for your confidence prob in image

### ISSUE THREAD

**Issue by moondra2017** [ecaradec/humble-yolo#4]
I can't seem to get the loss down below .40 despite playing around with the learning rates.
The confidence probs printed with each prob only get to around .2 ish if there is text.

How did you train your model so well.
I am running your code as is.

**Comment by off99555** [comment]
Yes. This code seems to be problematic. I observe that the `y_train` has `xy` values ranging from 0 to 2 (grid width, grid height). But this does not make sense because the required range should be between 0 to 1.
If you activate prediction by sigmoid you will get at most 1 so your minimum loss will never be close to 0 because your true value is 2.

It seems the problem starts at the dataset.
@ecaradec Did you publish the code with the correct version as shown in the article?

**Comment by ecaradec** [comment]
I agree, the code as it is doesn't seems to give as good result than when I wrote the article. I probably made a few changes for the articles, but it's been a while and I'm unsure what the problem is. I'll try to have a longer look at it.

y_train range is not the problem ihmo. The ranging goes from 0 to 2 on purpose. I'm adding an offset before computing loss with the following line. Look at code using the grid variable. If you only use xy_loss, this learn perfectly.

I think the loss not going to 0 is caused by the class prediction as these are two probabilities that are summed and there is little guaranties that it goes down to 0. You can experiment with removing the different loss function to see how the loss change.

It doesn't really matter that the loss goes down to 0 as long as it goes down in the right direction. 0 doesn't means the prediction is good either.

**Comment by off99555** [comment]
I run the code as is and I don't see the prediction is good at all. The thing is that the loss does not go lower than 0.4 like @moondra2017 said. You should check the code and make sure you are able to replicate the result when running the code "as is".

**Comment by goldengrisha** [comment]
@ecaradec, would mind writing what the version of python did you use?

**Comment by 0xlouis** [comment]
I agree.
A few weeks ago I tested the code, the results were not good.
I pointed out problems in the function that performs the loss calculation.
I rewrote the code to get better results (it's available in my fork).

**Comment by goldengrisha** [comment]
@0xlouis, you've rewritten everything.

**Comment by ecaradec** [comment]
I reduced the learning rate and made a small fix on the confidence score. I'm getting appropriate confidence score now : misplaced rect have low confidence and correctly place rect have great confidence. Give it a try if you want and let me know.

**Comment by goldengrisha** [comment]
@ecaradec , I've tested the accuracy:  
![image](https://user-images.githubusercontent.com/1469234/72210189-759fe600-34c0-11ea-9896-62791717ff35.png),

**Comment by ecaradec** [comment]
Default accuracy metrics is categorical accuracy. I don't think that it has an appropriate meaning here. What are you trying to measure ?

**Comment by goldengrisha** [comment]
@ecaradec ,  yes that's right, sorry it was my mistake, what metrics can I use to see the proper accuracy?

**Comment by ecaradec** [comment]
@goldengrisha I close the topic. Feel free to ask the question on a separate issue.
